src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleGn4SecurityHeaderAppender.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * (c) 2003 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license,
+ * available at the root application directory.
+ */
+package org.geonetwork.gn4forwarding;
+
+import static org.geonetwork.gn4forwarding.Gn4SecurityHeaderAppender.log1;
+import static org.springframework.web.servlet.function.HandlerFilterFunction.ofRequestProcessor;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.gateway.server.mvc.common.Shortcut;
+import org.springframework.http.HttpHeaders;
+import org.springframework.security.core.context.SecurityContextImpl;
+import org.springframework.security.core.userdetails.User;
+import org.springframework.web.servlet.function.HandlerFilterFunction;
+import org.springframework.web.servlet.function.ServerRequest;
+import org.springframework.web.servlet.function.ServerResponse;
+
+/**
+ * See Gn4SecurityHeaderAppender, which is a more complicated version of this filter.
+ *
+ * <p>This ONLY extracts the UserName (no rights or anything else) and sends that to GN4. GN4 needs to be setup to NOT
+ * update profiles or groups. GN4 will get the rights from the DB.
+ */
+public class SimpleGn4SecurityHeaderAppender {
+    private static final Object LOCK = new Object();
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(Gn4SecurityHeaderAppender.class);
+    // debugging?  Set this to true.  This makes each request happen sequencially.
+    // This is useful for debugging (all logging together)
+    // but, has a performance impact.
+    private static final boolean oneRequestAtATime = true;
+
+    /**
+     * main entry for the GN5->GN4 security token. JUST USERNAME
+     *
+     * @param request incomming gn4 request
+     * @param gn4AuthHeaderName name of the header to set with security info
+     * @return new request that has the gn4AuthHeaderName (with json security info - JUST USERNAME)
+     */
+    public static ServerRequest addSimpleGn4SecurityHeader_impl(ServerRequest request, String gn4AuthHeaderName) {
+        log1(request);
+
+        // remove the output header (or its a MAJOR security problem)
+        var changedRequest = ServerRequest.from(request).headers(new Consumer<HttpHeaders>() {
+            // security - remove this incoming header or it will be trusted in GN4!!
+            @Override
+            public void accept(HttpHeaders httpHeaders) {
+                httpHeaders.remove(gn4AuthHeaderName);
+            }
+        });
+
+        var user = getUser(request);
+        if (user == null) {
+            return null;
```

Comment:
```suggestion
            return changedRequest.build();
```

we must return a request here. Not 100% sure if we return this or the incoming request?

Reply:
good catch!